// atom-package-manager/lib/test.js
(function() {
  var Command, Test, fs, optimist, path, temp,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  optimist = require('optimist');

  temp = require('temp');

  Command = require('./command');

  fs = require('./fs');

  module.exports = Test = (function(_super) {
    __extends(Test, _super);

    function Test() {
      return Test.__super__.constructor.apply(this, arguments);
    }

    Test.commandNames = ['test'];

    Test.prototype.parseOptions = function(argv) {
      var options;
      options = optimist(argv);
      options.usage("Usage:\n  apm test\n\nRuns the package's tests contained within the spec directory (relative\nto the current working directory).");
      options.alias('h', 'help').describe('help', 'Print this usage message');
      return options.alias('p', 'path').string('path').describe('path', 'Path to atom command');
    };

    Test.prototype.showHelp = function(argv) {
      return this.parseOptions(argv).showHelp();
    };

    Test.prototype.getChocolateyAtomPath = function() {
      var atomCommand;
      if (process.env.CHOCOLATEYINSTALL) {
        atomCommand = path.join(process.env.CHOCOLATEYINSTALL, 'bin', 'atom.exe');
        if (fs.existsSync(atomCommand)) {
          return atomCommand;
        }
      }
      if (process.env.ALLUSERSPROFILE) {
        atomCommand = path.join(process.env.ALLUSERSPROFILE, 'chocolatey', 'bin', 'atom.exe');
        if (fs.existsSync(atomCommand)) {
          return atomCommand;
        }
      }
      return null;
    };

    Test.prototype.run = function(options) {
      var atomCommand, callback, cmd, cmdArgs, cmdOptions, env, logFile, logFilePath, packagePath, testArgs;
      callback = options.callback;
      options = this.parseOptions(options.commandArgs);
      env = process.env;
      if (options.argv.path) {
        atomCommand = options.argv.path;
      } else if (process.platform === 'win32') {
        atomCommand = this.getChocolateyAtomPath();
      }
      if (!fs.existsSync(atomCommand)) {
        atomCommand = 'atom';
      }
      packagePath = process.cwd();
      testArgs = ['--dev', '--test', "--spec-directory=" + (path.join(packagePath, 'spec'))];
      if (process.platform === 'win32') {
        testArgs.unshift('--shimgen-waitforexit');
        logFile = temp.openSync({
          suffix: '.log',
          prefix: "" + (path.basename(packagePath)) + "-"
        });
        fs.closeSync(logFile.fd);
        logFilePath = logFile.path;
        testArgs.push("--log-file=" + logFilePath);
        cmdArgs = testArgs.map(function(arg) {
          return "\"" + (arg.replace(/"/g, '\\"')) + "\"";
        });
        cmdArgs.unshift("\"" + atomCommand + "\"");
        cmdArgs = ['/s', '/c', "\"" + (cmdArgs.join(' ')) + "\""];
        cmdOptions = {
          env: env,
          windowsVerbatimArguments: true
        };
        cmd = process.env.comspec || 'cmd.exe';
        return this.spawn(cmd, cmdArgs, cmdOptions, function(code) {
          var loggedOutput;
          try {
            loggedOutput = fs.readFileSync(logFilePath, 'utf8');
            if (loggedOutput) {
              process.stdout.write("" + loggedOutput + "\n");
            }
          } catch (_error) {}
          if (code === 0) {
            process.stdout.write('Tests passed\n'.green);
            return callback();
          } else {
            return callback('Tests failed');
          }
        });
      } else {
        return this.spawn(atomCommand, testArgs, {
          env: env,
          streaming: true
        }, function(code) {
          if (code === 0) {
            process.stdout.write('Tests passed\n'.green);
            return callback();
          } else {
            return callback('Tests failed');
          }
        });
      }
    };

    return Test;

  })(Command);

}).call(this);
